core/src/main/java/org/apache/iceberg/PartitionSpecUpdate.java

```diff
@@ -0,0 +1,214 @@
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addYearField(String sourceName) {
+    newSpecBuilder.year(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addMonthField(String sourceName, String targetName) {
+    newSpecBuilder.month(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addMonthField(String sourceName) {
+    newSpecBuilder.month(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addDayField(String sourceName, String targetName) {
+    newSpecBuilder.day(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addDayField(String sourceName) {
+    newSpecBuilder.day(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addHourField(String sourceName, String targetName) {
+    newSpecBuilder.hour(sourceName, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addHourField(String sourceName) {
+    newSpecBuilder.hour(sourceName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addBucketField(String sourceName, int numBuckets, String targetName) {
+    newSpecBuilder.bucket(sourceName, numBuckets, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addBucketField(String sourceName, int numBuckets) {
+    newSpecBuilder.bucket(sourceName, numBuckets);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width, String targetName) {
+    newSpecBuilder.truncate(sourceName, width, targetName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addTruncateField(String sourceName, int width) {
+    newSpecBuilder.truncate(sourceName, width);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addField(int sourceId, String name, String transform) {
+    newSpecBuilder.add(sourceId, name, transform);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec addField(int sourceId, String name, Transform<?, ?> transform) {
+    return addField(sourceId, name, transform.toString());
+  }
+
+  @Override
+  public UpdatePartitionSpec renameField(String name, String newName) {
+    newSpecBuilder.rename(name, newName);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec removeField(String name) {
+    newSpecBuilder.remove(name, base.formatVersion() == 1);
+    return this;
+  }
+
+  @Override
+  public UpdatePartitionSpec replaceField(String name, String transform) {
+    int sourceId = newSpecBuilder.remove(name, base.formatVersion() == 1);
+    return addField(sourceId, name, transform);
+  }
+
+  @Override
+  public UpdatePartitionSpec replaceField(String name, Transform<?, ?> transform) {
+    return replaceField(name, transform.toString());
+  }
+
+  @Override
+  public void commit() {
+    TableMetadata update = base.updatePartitionSpec(freshSpecFieldIds(apply()));
```

Comment:
Should we also track `lastAssignedPartitionId` in metadata?

Reply:
`lastAssignedPartitionId` is kept inside each partition spec. metadata's `lastAssignedPartitionId ` is only used here before committing. So we can lazily compute it while looping over the partition spec history. So it is unnecessary to add an extra field into Table metadata.

Reply:
I think you need to add a test case where the current spec's highest ID is lower than another spec's highest ID. We can't reuse IDs if the user removes a field, commits, and starts a new change to add a new field.

Reply:
Thanks for the comment. Yep, it handles that situation as the computation loops over all the partition specs in the history. I will add a test for this case.